### DISCUSSION TITLE
[help] Load balancing to prevent workers being idel with tar_make_future

### DISCUSSION THREAD

**Discussion by mem48** [ropensci/targets#1143] [Help] [answered]
### Help

- [X] I understand and agree to <https://books.ropensci.org/targets/help.html>.

### Description

I have a workflow that includes some long-running targets (several hours) and many quick targets (a few minutes). There is a lot of parallelism in the workflow, so I'm using `tar_make_future` with `workers = 4`.  My hope was that while one worker worked on a slow task the other three would do many of the quick tasks.

But this does not appear to be the case. Watching the process with `tar_watch` suggests that when targets are handed out to workers each worker gets a single target and new targets are not allocated until all 4 workers are finished. This means that 3 workers sit idle for most of the time even though there are plenty of queued targets. 

I know that `future` is capable of load balancing and allocating new jobs to workers as soon as the previous one is completed. But I can't find a way to make this happen with targets.

**Comment by wlandau** [comment]
`tar_make_future()` creates a fresh `future()` for each target. For `future` backends like the ones in `future.batchtools`, this prevents idling. If your backend is a custom PSOCK cluster, there is a risk of idling, but `targets` and loops over the futures to try and keep at least 4 running at a time (for `workers = 4`).

That said, I recommend switching to `crew` and setting `seconds_idle` to something useful: https://books.ropensci.org/targets/crew.html

**Reply by mem48** [reply]
Could this be a Windows issue? I think Windows uses PSOCK by default?

**Reply by wlandau** [reply]
The default `future` plan is `sequential`.

Please have a look at https://books.ropensci.org/targets/help.html. Reprexes are important for diagnosing issues like these.

**Comment by mem48** [comment]
Reprex

```
library(targets)

sleep_run <- function(x) {
    Sys.sleep(x)
    x
}

future::plan(future::multisession, workers = 4)

list(tar_target(fast_job1, {
    sleep_run(1)
}), tar_target(slow_job1, {
    length(fast_job1)
    sleep_run(30)
}), tar_target(fast_job2, {
    length(fast_job1)
    sleep_run(3)
}), tar_target(fast_job3, {
    length(fast_job2)
    sleep_run(3)
}), tar_target(fast_job4, {
    length(fast_job3)
    sleep_run(3)
}), tar_target(fast_job5, {
    length(fast_job4)
    sleep_run(3)
}), tar_target(fast_job6, {
    length(fast_job5)
    sleep_run(3)
}))

```
This should run in 31 seconds, but actually take 48 seconds.

**Comment by mem48** [comment, marked as answer]
I think I've found the solution

```
library(targets)
library(future)
library(future.callr)


sleep_run <- function(x) {
    Sys.sleep(x)
    x
}

plan(callr)

list(tar_target(fast_job1, {
    sleep_run(1)
}), tar_target(slow_job1, {
    length(fast_job1)
    sleep_run(30)
}), tar_target(fast_job2, {
    length(fast_job1)
    sleep_run(3)
}), tar_target(fast_job3, {
    length(fast_job2)
    sleep_run(3)
}), tar_target(fast_job4, {
    length(fast_job3)
    sleep_run(3)
}), tar_target(fast_job5, {
    length(fast_job4)
    sleep_run(3)
}), tar_target(fast_job6, {
    length(fast_job5)
    sleep_run(3)
}))

```

Takes 32 seconds

Not sure why `future::plan(future::multisession, workers = 4)` doesn't work
